### DISCUSSION TITLE
Can I run docker in codespaces?

### DISCUSSION THREAD

**Discussion by khantseithu** [community/community#48003] [Codespaces] [answered]
### Select Topic Area

Question

### Body

I'm doing a course from frontend masters about sql which require to setup docker. Unfortunately, my pc is really old and only have 4GB pf ram. So, I was wandering if I could follow along with the course using codespaces. If possible, I'd really appreciate your help setting up my codespace since I don't know anything about docker. Here is the docker setup from the course: https://sql.holt.courses/lessons/welcome/docker

Two containers I'll need:
https://hub.docker.com/_/postgres/
https://hub.docker.com/r/btholt/complete-intro-to-sql

**Comment by garrettw** [comment, marked as answer]
Just as a brief answer, yes you can. You don't actually need to run Docker *inside* of your codespace, as it is already a Docker container of its own. What you should do is define a [devcontainer config](https://containers.dev/) along with a [docker-compose config](https://docs.docker.com/compose/compose-file/), and then just create a new codespace not from scratch, but from your devcontainer config.

**Reply by practical-dreamer** [reply]
how tho

**Reply by khantseithu** [reply]
you can just run it, docker and docker compose is already installed in every codespaces by default.

**Reply by kleenkanteen** [reply]
This helped me the most in setting up the devcontainer to be created from a Dockerfile: [docs](https://docs.github.com/en/codespaces/setting-up-your-project-for-codespaces/adding-a-dev-container-configuration/introduction-to-dev-containers)

**Comment by babski123** [comment]
Yes, this is possible. I just tried it today.
Just treat the codespace terminal as if it is your computer's terminal.

**Comment by Gamor-Miracle-Kwasi** [comment]
Absolutely, you can get Docker going in GitHub Codespaces. However, it’s worth mentioning that it doesn’t come pre-installed, so a bit of setup is required. Essentially, to enable Docker functionality, you'll need to put together a .devcontainer configuration in your repository.

This involves adding the installation of Docker and mounting the Docker socket (so that) commands related to Docker can run inside your Codespace environment. This way, you’re able to manage containers, build images for Docker (and work with other applications), all from within your cloud-based development space.

Yet keep in mind that running Docker here typically depends on what's called "Docker-in-Docker" or DinD. Proper resource management tends to be very important too since it helps prevent issues like excessive use of CPU and memory which could otherwise slow things down significantly.

**Reply by stevekm** [reply]
> mounting the Docker socket (so that) commands related to Docker can run inside your Codespace environment.

How do you do this?

Also it seems like it would be unsafe to be using the host's Docker socket to run Docker inside the CodeSpaces instance. How do you know that the host's socket has not been compromised?

**Reply by arvesv** [reply]
I just tried this. If you are running your Codespace on a specific image, then the  "Docker in Docker"  feature may not have been installed. But you can add the feature in devcontainer.json.  I used the 

`	"ghcr.io/devcontainers/features/docker-in-docker:2": {}
`
and then it just worked

**Reply by davidxia** [reply]
I think @khantseithu's original question is more about how to use Github Codespaces. But for anyone who stumbled here like me who wants to use Docker inside a Codespace, @Gamor-Miracle-Kwasi's answer is one approach that works.

> How do you do this?

Below works for me. I'm able to run `docker` commands like `docker ps` and all other types of docker operations.

```json
{
  "name": "input-plane",
  ...
  "mounts": [
    "source=/var/run/docker.sock,target=/var/run/docker.sock,type=bind"
  ]
}
```

Right now `docker version` shows.

```console
[root@codespaces-e63915]# docker version
Client: Docker Engine - Community
 Version:           28.4.0
 API version:       1.43 (downgraded from 1.51)
 Go version:        go1.24.7
 Git commit:        d8eb465
 Built:             Wed Sep  3 20:59:57 2025
 OS/Arch:           linux/amd64
 Context:           default

Server:
 Engine:
  Version:          24.0.9-2
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.21.9
  Git commit:       fca702de7f71362c8d103073c7e4a1d0a467fadd
  Built:            Thu Feb  1 00:12:23 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.7.27-1
  GitCommit:        05044ec0a9a75232cad458027ca83437aae3f4da
 runc:
  Version:          1.1.15-1
  GitCommit:        bc20cb4497af9af01bea4a8044f1678ffca2745c
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

> Also it seems like it would be unsafe to be using the host's Docker socket to run Docker inside the CodeSpaces instance. How do you know that the host's socket has not been compromised?

I think each Codespace gets its own dedicated VM. So this shouldn't be an issue.

**Reply by 5078086729442496Mahmoud** [reply]
3928

**Comment by star3am** [comment]
Yes you can, there is the Docker in Docker way, but I needed something a little different, I run a doecker daemon inside the codespace, see https://github.com/star3am/hashiqube
Here's the Dev container file https://github.com/star3am/hashiqube/tree/master/.devcontainer and that points to the docker-compose file https://github.com/star3am/hashiqube/blob/master/docker-compose.yml specifically see the tmpfs volumes and the docker privileged and cgroups host. Here I install the docker daemon inside the container https://github.com/star3am/hashiqube/blob/master/docker/docker.sh

[a comment by an automated account is not shown]
